### PR TITLE
Merging database extent and ranges with datacube-core

### LINKED ISSUE #366: Quickly Determine Extents of a Product
It is extremely slow to query the full available extents for a large Product in a Data Cube Index.

This has been frustrating for several developers, and has led to multiple implementations of precomputing and caching this information. Details in the code of each of the following projects:

   - [DEA Dashboard (cubedash)](https://github.com/data-cube/dea-dashboard)
   - [Data Cube Web Map Service](https://github.com/opendatacube/datacube-wms)
   - [WCS Server](https://github.com/ceos-seo/django-datacube-wcs)

This slowness also affects the pre-flight job tasking which is done 

A simple solution would be an extra database table containing extents and stats about each product, with an offline job which is run regularly to keep it up to date.

The easiest way to implement this would be to take the code from one of the above projects, and include it in `datacube-core`, along with a command to update the table.

## Outstanding Questions
* Can the code from one of these projects be used almost directly, or will new code be required?
* Should the new table store a simple bounding box, or use PostGIS powers to calculate a more accurate vector extent?
* We don't have a robust _migrations_ system for upgrade the Database Schema, it's handled by some bespoke code. Is it worth improving this at the same time, or hacking in something which runs automatically, or having a manually run command to enable this *Extents* extension.

### PR DESCRIPTION
**Pull request by ashoka1234**
### Reason for this pull request
Extents polygons and bounding boxes are computed for products as a whole as well as for specific time spans. Often list of extents require to be computed for regular intervals which can be yearly, monthly, or daily. Some applications require time bounds as well as spatial bounds product-wide. 

### Proposed changes
Three database tables are proposed to maintain spatial extents, spatial bounds, and time bounds: extent_meta, extent, and ranges. Extent_meta contain meta information regarding spatial extents stored in the extent table. Extent table stores exact spatial boundary of a product for a specific period. The periods are specified with python pandas style timestamp (start) and offset from the start. Offsets are pandas style offset alias strings such as `1M` indicating a month, `1Y` indicating a year etc. The ranges table stores product-wide information such as time bounds and spatial bounds. 

The database-facing retrieval functions are integrated in `datacube.drivers.postgres` module while user-facing retrieval API is located in `ProductResource` class. The schema and migration code are located in `datacube.drivers.postgres` module. The upload scripts are configurable as in `db_extent_cfg.yaml`. The extent and ranges upload script is located in `db_extent_update.py` which rely on a upload API located in `_extent_upload.py`. The uploads are designed to be fast using multiprocessing with task splits across time. Further upload scripts verify what is already done as per information available in the extent_meta table and computes only extents that are not yet computed. Similarly ranges uploads only attempt to see datasets that are not yet seen. So repeated runs of the upload script are fast.  

This pull request replaces #430 

 - [ ] Closes #366 
 - [ ] Tests added / passed
 - [ ] Fully documented, including `docs/about/whats_new.rst` for all changes

### CODE REVIEW

**Comment by omad** [comment]
There's a lot of good code in here Ashoka, but there's a few things we need to improve before merging.
- Turn the extents generation tool into an actual tool
- Move the example code out and into documentation
- Include the docs as part of the real documentation
- Add and improve the tests

File: datacube/scripts/product_extent.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+from datacube.db_extent import ExtentUpload, parse_time
+from datacube.drivers.postgres import PostgresDb
+from datacube.index.index import Index
+from datacube.ui import click as ui
+from datacube.ui.click import cli
+
+_LOG = logging.getLogger('datacube-md-type')
+
+
+@cli.group(name='product_extent', help='Product Extents Tools')
```
Comment (Kirill888) [inline datacube/scripts/product_extent.py]: Can we just call it `extent`, having `_` in the subcommand feels odd, also we already have just `product` subcommand.

Comment (omad) [reply datacube/scripts/product_extent.py]: Good idea, we're still working on making it feel integrated.

**Comment by jeremyh** [comment]
Thanks Ashoka, it looks good so far. I'm not sure when to add comments as it still seems to be a moving target, and I know Damien already gave you some guidance on integrating it more with the existing code. 

These are a few rough, vague and higher-level notes:

- It still needs some thought to improve how it interacts with drivers: eg. the `extent_upload`is a top-level module, living outside the drivers, but is completely tied to the PostgresDb driver. It's also interacting both with the lowest level api (PostgresDB) and highest level (Datacube) in the same file. We should separate the driver-specific and driver-agnostic parts. Any usage of the high-level api (`Datacube`) prevents it from using these apis due to circular imports (Perhaps replace `dc.find_datasets_lazy` with lower-level `index.datasets.search`?)

- Damien probably already mentioned it, but I'd remove the parts that pass around individual config settings (hostname/password/etc) and use the existing datacube config handling instead. We then also won't need to hard-code dev postgres settings, as people already will have their (dev) settings in `datacube.conf` files.

- The code is creating its own sqlalchemy engine: this might need some cleanup, and if possible use the existing engine setup in PostgresDb (or at least share code with it). It's missing the application name and connection/pool renewal settings and it would be nice not to have to make future connection/alchemy changes in multiple places.

- Currently the pooling/connection handling is disabled, leading to an expensive connection setup (and validation) for almost every query. I ran a test of one month of one product on my laptop and it made ~450 individual connections and ~9400 queries.  Running it a second time (to incrementally update) took significantly longer (I killed it after a few minutes) -- these might be related to #491, though?

- Allow the `extent update` command-line command to take a list of products, and maybe an `--all` option? Most of the time I want to update all of the extents on a system.

- For implementing the tests, there's some existing pytest fixtures that you can use  to test against a real database: have a look at [integration_tests/index/test_search.py](https://github.com/opendatacube/datacube-core/blob/develop/integration_tests/index/test_search.py) as an example that uses them.

- It also might be nice to add a few unit tests for the simple methods with `compute`/`util` in the name. A simple doctest would work on many of them: here's an existing example: (pytest is configured to scan all docstrings for them: [example](https://github.com/opendatacube/datacube-core/blob/d951815a6f8b8bda1e95e7691b00360110358ba5/datacube/model/__init__.py#L681))

- Add a test for updating extents of a product with no datasets, and one with no `time` and/or `lat/lon` fields (eg. satellite_telemetry_data). These are throwing errors.

- It's worth reviewing the API parameter and return types. 
	- Avoid returning sqlalchemy-specific row records: the use of sqlalchemy by the driver should be an implementation detail, not exposed to users. You can wrap a row with `dict()` before returning, or even better use a dataclass/nametuple/etc.
	- The `ranges()` function especially needs improvement: it returns an sqlalchemy row record, for which "bounds" JSONB is returned as a raw string. So I call `bounds()` and then`json.loads(row["bounds"])` to get a dict of top/bottom/left/right fields. (sqlalchemy usually returns jsonb as a dict, but the code is using table reflection rather than your table definitions, so it's not identifying the postgres types properly).
	- Most of our other index apis use a `datacube.model.Range` object for a range, and elsewhere we  sometimes use a two-element tuple. I'd use one of these rather than adding a third representation (separate "time_min" and "time_max" fields)

- The ability to query arbitrary extents is neat! 


A few points I'm unsure about — take with a grain of salt!

- The use of the uuid for extent lookup. Note that it's larger in size than using the fields themselves as a composite (dataset_type_ref is 2 bytes, start date is 4 bytes, offset is char 1 (or 2?) bytes == 8 bytes, vs UUID 16 bytes). It has the downside of being unreadable/obtuse by developers when looking at the table, and is difficult to query/compute in SQL. It presumably may index better though, avoiding a composite index? I haven't tested.

- Storing/using pandas offset aliases as the basis of the api. This seems extremely flexible at the top api level, but I'm less sure about storing them at the low level. It's doing exact string matching on "1M" and "1Y", but in theory users could supply lots of other variations ("1m", "1m0d" ... I think?). These strings are stored in Postgres but can't be understood by Postgres  queries themselves (unless we assume uses only every pass exactly "1M" and "1Y"). If we *effectively* always expect month/days, perhaps we could just use a simple enum? Or if flexibility is needed, we could look at a more portable/convertible representation for it: Postgres supports iso durations and/or has its own duration type.

- The range `bounds` column is defined as `jsonb` but is always four numbers: left/right/up/down. In lieu of adding Postgis as a dependency, the built-in Postgres `box` type or even a four-element array are simpler and much smaller on disk, assuming we're not using the field dynamically(?)

---

Please feel free to come and discuss any of these, especially testing or the drivers/layering!

**Comment by jeremyh** [comment]
Also, I'd open the floor for team discussions on the names of things. Damien may be better at this than I am.

- The `db_extent` module name. It's inconsistent with our other modules, and "db_" is unnecessary. Maybe just `extent`, or remove it entirely if it's being merged into the index package.

- Table/entity names:

| Name | Current Python Comment | Maybe rename to? | 
| :--- | :--- | :--- |
| `extent` | "The spatial extent geometry for various time periods. Refer to extent_meta table to obtain length of time period." | `extent_slice` |
| `extent_meta` |  "Metadata for spatial extents." | `extent` |
| `ranges` | "Time min/max and rectangular spatial bounds for products" | `product_range` (or `dataset_type_range` for table consistency: it's probably less confusing until we rename the other one) |

- Some of the api function name could be reviewed. `get_extent()` and `get_extent_direct()` could be more distinguishable from each other?

**Comment by ashoka1234** [comment]
Thanks Jeremy. Will get onto those one by one after discussing with Damien.

**Comment by ashoka1234** [comment]
The stuff in _extent_upload is not merged yet and we haven't given priority as far as merge with existing code is concerned. The retrieval stuff are now more integrated with exiting code so does not use any driver specific details as far as functions that are merged in ProductResource concerned. Also we no longer use uuid's for retrieval  so I guess it could just be removed altogether or replaced with a int?

**Comment by omad** [comment]
We still want a feature like this, but there's too many outstanding issues with this to be able to merge.

**Pull request closed without merging**